From 3361fa23448c1b165b89e3a20f06e48458251655 Mon Sep 17 00:00:00 2001
From: Kent Tamura <tkent@chromium.org>
Date: Tue, 10 Sep 2024 06:50:41 +0000
Subject: [PATCH] Editing: Do not crash on locking a block including a caret

Bug: 40702987, 364687924
Change-Id: Ia0f8d7b687b9fac1395fdcb0b00b937b5764e8c8
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5848029
Commit-Queue: Koji Ishii <kojii@chromium.org>
Reviewed-by: Koji Ishii <kojii@chromium.org>
Auto-Submit: Kent Tamura <tkent@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1353161}
---

--- a/third_party/blink/renderer/core/layout/inline/inline_caret_position.cc
+++ b/third_party/blink/renderer/core/layout/inline/inline_caret_position.cc
@@ -453,9 +453,8 @@
 
   const OffsetMapping* const mapping = InlineNode::GetOffsetMapping(context);
   if (!mapping) {
-    // TODO(yosin): We should find when we reach here[1].
-    // [1] http://crbug.com/1100481
-    DUMP_WILL_BE_NOTREACHED() << context;
+    // A block containing the position might be display-locked.
+    // See editing/caret/caret-display-locked-crash.html
     return InlineCaretPosition();
   }
   const std::optional<unsigned> maybe_offset =
--- /dev/null
+++ b/third_party/blink/web_tests/editing/caret/caret-display-locked-crash.html
@@ -0,0 +1,44 @@
+<!DOCTYPE html>
+<html>
+<head>
+<script src="../../resources/testharness.js"></script>
+<script src="../../resources/testharnessreport.js"></script>
+<title>`content-visibility` crash in `[contenteditable]` after focus</title>
+<style>
+p {
+  height: 600px;
+}
+#p0 {
+  content-visibility: auto;
+}
+</style>
+</head>
+<body>
+<div contenteditable="true">
+<p id="p0">paragraph 0</p>
+<p>paragraph 1</p>
+<p>paragraph 2</p>
+<p>paragraph 3</p>
+<p>paragraph 4</p>
+<p>paragraph 5</p>
+<p>paragraph 6</p>
+</div>
+<script>
+async function animationFrame() {
+  return new Promise(resolve => {
+    requestAnimationFrame(resolve);
+  });
+}
+
+promise_test(async t => {
+  document.querySelector("div").focus();
+  await animationFrame();
+  await animationFrame();
+  // Locks #p0, which contains the caret.
+  window.scrollBy(0, 2000);
+  await animationFrame();
+  await animationFrame();
+}, 'No crash on hiding a c-v:auto block containing a text caret.');
+</script>
+</body>
+</html>
